In [1]:
import numpy as np
import pandas as pd

In [2]:
orders = pd.DataFrame([
    {"дата": "2026-09-01", "місто": "Івано-Франківськ", "товар": "Товар 7", "ціна": 760, "кількість": 3},
    {"дата": "2026-09-02", "місто": " Івано-Франківськ ", "товар": "Товар 7", "ціна": "760 грн", "кількість": 4},
    {"дата": "2026-09-03", "місто": "ІВАНО-ФРАНКІВСЬК", "товар": "Товар 7", "ціна": 1520, "кількість": np.nan},
    {"дата": "2026-09-04", "місто": "Івано-Франківськ", "товар": "Товар 7", "ціна": "760 грн", "кількість": 2},
    {"дата": "2026-09-05", "місто": "Івано-Франківськ", "товар": "Товар 7", "ціна": 7600, "кількість": 5},
    {"дата": "2026-09-06", "місто": "Івано-Франківськ", "товар": "Товар 7", "ціна": 700, "кількість": np.nan},
    {"дата": "2026-09-07", "місто": "Івано-Франківськ", "товар": "Товар 7", "ціна": 780, "кількість": 3},
    {"дата": "2026-09-07", "місто": "Івано-Франківськ", "товар": "Товар 7", "ціна": 780, "кількість": 3},
])

orders

,дата,місто,товар,ціна,кількість
0,2026-09-01,Івано-Франківськ,Товар 7,760,3.0
1,2026-09-02,Івано-Франківськ,Товар 7,760 грн,4.0
2,2026-09-03,ІВАНО-ФРАНКІВСЬК,Товар 7,1520,NaN
3,2026-09-04,Івано-Франківськ,Товар 7,760 грн,2.0
4,2026-09-05,Івано-Франківськ,Товар 7,7600,5.0
5,2026-09-06,Івано-Франківськ,Товар 7,700,NaN
6,2026-09-07,Івано-Франківськ,Товар 7,780,3.0
7,2026-09-07,Івано-Франківськ,Товар 7,780,3.0


In [3]:
orders.isna().sum()

дата         0
місто        0
товар        0
ціна         0
кількість    2
dtype: int64

In [4]:
orders["кількість"].median()

np.float64(3.0)

In [5]:
orders["кількість"] = orders["кількість"].fillna(orders["кількість"].median())

orders

,дата,місто,товар,ціна,кількість
0,2026-09-01,Івано-Франківськ,Товар 7,760,3.0
1,2026-09-02,Івано-Франківськ,Товар 7,760 грн,4.0
2,2026-09-03,ІВАНО-ФРАНКІВСЬК,Товар 7,1520,3.0
3,2026-09-04,Івано-Франківськ,Товар 7,760 грн,2.0
4,2026-09-05,Івано-Франківськ,Товар 7,7600,5.0
5,2026-09-06,Івано-Франківськ,Товар 7,700,3.0
6,2026-09-07,Івано-Франківськ,Товар 7,780,3.0
7,2026-09-07,Івано-Франківськ,Товар 7,780,3.0


In [6]:
orders.duplicated().sum()

np.int64(1)

In [7]:
orders.duplicated(subset=["дата", "товар"]).sum()

np.int64(1)

In [8]:
orders = orders.drop_duplicates(subset=["дата", "товар"])

orders

,дата,місто,товар,ціна,кількість
0,2026-09-01,Івано-Франківськ,Товар 7,760,3.0
1,2026-09-02,Івано-Франківськ,Товар 7,760 грн,4.0
2,2026-09-03,ІВАНО-ФРАНКІВСЬК,Товар 7,1520,3.0
3,2026-09-04,Івано-Франківськ,Товар 7,760 грн,2.0
4,2026-09-05,Івано-Франківськ,Товар 7,7600,5.0
5,2026-09-06,Івано-Франківськ,Товар 7,700,3.0
6,2026-09-07,Івано-Франківськ,Товар 7,780,3.0


In [9]:
orders["ціна"] = (
    orders["ціна"]
    .astype(str)
    .str.replace(" грн", "", regex=False)
    .astype(float)
)

orders

,дата,місто,товар,ціна,кількість
0,2026-09-01,Івано-Франківськ,Товар 7,760.0,3.0
1,2026-09-02,Івано-Франківськ,Товар 7,760.0,4.0
2,2026-09-03,ІВАНО-ФРАНКІВСЬК,Товар 7,1520.0,3.0
3,2026-09-04,Івано-Франківськ,Товар 7,760.0,2.0
4,2026-09-05,Івано-Франківськ,Товар 7,7600.0,5.0
5,2026-09-06,Івано-Франківськ,Товар 7,700.0,3.0
6,2026-09-07,Івано-Франківськ,Товар 7,780.0,3.0


In [10]:
orders["місто"].unique()

<StringArray>
['Івано-Франківськ', ' Івано-Франківськ ', 'ІВАНО-ФРАНКІВСЬК']
Length: 3, dtype: str

In [11]:
orders["місто"] = (
    orders["місто"]
    .str.strip()
    .str.lower()
    .replace({"івано-франківськ": "Івано-Франківськ"})
)

orders["місто"].unique()

<StringArray>
['Івано-Франківськ']
Length: 1, dtype: str

In [12]:
Q1 = orders["ціна"].quantile(0.25)
Q3 = orders["ціна"].quantile(0.75)

IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print("Q1 =", Q1)
print("Q3 =", Q3)
print("IQR =", IQR)
print("Нижня межа =", lower_bound)
print("Верхня межа =", upper_bound)

Q1 = 760.0
Q3 = 1150.0
IQR = 390.0
Нижня межа = 175.0
Верхня межа = 1735.0


In [13]:
outliers = orders[
    (orders["ціна"] < lower_bound) |
    (orders["ціна"] > upper_bound)
]

outliers

,дата,місто,товар,ціна,кількість
4,2026-09-05,Івано-Франківськ,Товар 7,7600.0,5.0


In [14]:
orders.loc[orders["ціна"] == 7600, "ціна"] = 760

orders

,дата,місто,товар,ціна,кількість
0,2026-09-01,Івано-Франківськ,Товар 7,760.0,3.0
1,2026-09-02,Івано-Франківськ,Товар 7,760.0,4.0
2,2026-09-03,Івано-Франківськ,Товар 7,1520.0,3.0
3,2026-09-04,Івано-Франківськ,Товар 7,760.0,2.0
4,2026-09-05,Івано-Франківськ,Товар 7,760.0,5.0
5,2026-09-06,Івано-Франківськ,Товар 7,700.0,3.0
6,2026-09-07,Івано-Франківськ,Товар 7,780.0,3.0


In [15]:
print("Пропуски:")
print(orders.isna().sum())

print("\nКількість дублікатів:", orders.duplicated().sum())

print("\nТипи даних:")
print(orders.dtypes)

print("\nУнікальні міста:")
print(orders["місто"].unique())

print("\nЦіни:")
print(orders["ціна"].unique())

Пропуски:
дата         0
місто        0
товар        0
ціна         0
кількість    0
dtype: int64

Кількість дублікатів: 0

Типи даних:
дата             str
місто            str
товар            str
ціна         float64
кількість    float64
dtype: object

Унікальні міста:
<StringArray>
['Івано-Франківськ']
Length: 1, dtype: str

Ціни:
[ 760. 1520.  700.  780.]


### 1. Чому для заповнення пропусків у `кількість` було використано медіану, а не середнє?

У наборі даних лише 8 початкових спостережень, тому окремі великі або малі значення можуть суттєво впливати на середнє арифметичне. Медіана є більш стійкою до таких значень, тому для цього невеликого набору даних вона краще підходить для заповнення пропусків. Медіана кількості дорівнює 3, тому пропущені значення були замінені на 3.

### 2. Чим відрізняється `duplicated()` без `subset` від `duplicated()` із `subset`?

`duplicated()` без параметра `subset` перевіряє повний збіг усіх значень у рядках. Якщо всі стовпці однакові, рядок вважається дублікатом.

`duplicated(subset=["дата", "товар"])` перевіряє дублікати лише за вибраними стовпцями. У цій роботі дата та товар були використані як поля, що визначають унікальне замовлення.

### 3. Чому не можна автоматично видаляти всі значення, які виходять за межі IQR?

Значення за межами IQR не обов'язково є помилками. Вони можуть бути реальними та важливими спостереженнями. Тому перед видаленням викиду потрібно перевірити контекст даних.

У цьому наборі значення `7600 грн` було визнано помилкою введення, оскільки базова ціна для варіанта становить `760 грн`, значення `7600` рівно у 10 разів більше, а також перевищує верхню межу IQR `1735 грн`. Ймовірно, під час введення було додано зайвий нуль, тому значення було виправлено на `760 грн`.

### 4. Результат очищення даних

Після очищення у наборі залишилося 7 унікальних рядків. Пропущені значення кількості замінено медіаною, дубль видалено, ціни перетворено на числовий тип, назви міста нормалізовано, а аномальну ціну виправлено.
